File: Project2/Baitoan4/tree/cpp/graph_based_representation.cpp
#include <vector>
#include <iostream>
#include <fstream>
#include <string>

extern const int NIL_VALUE; // Use the constant from array_of_parents.cpp

/**
 * @brief Structure to represent graph-based representation for tree
 */
struct GraphBasedRepresentation {
    std::vector<std::vector<int>> adjacencyData;
    int numberOfNodes;
    int rootNode;
    
    /**
     * @brief Default constructor
     */
    GraphBasedRepresentation() : numberOfNodes(0), rootNode(NIL_VALUE) {}
    
    /**
     * @brief Constructor with node count
     * @param nodeCount Number of nodes in the tree
     */
    GraphBasedRepresentation(int nodeCount) : numberOfNodes(nodeCount), rootNode(NIL_VALUE) {
        adjacencyData.resize(nodeCount);
    }
};

/**
 * @brief Read graph-based representation from input file
 * @param fileName Name of input file
 * @return GraphBasedRepresentation structure containing the tree data
 */
GraphBasedRepresentation readGraphBasedRepresentationFromFile(const std::string& fileName) {
    std::ifstream inputFile(fileName);
    if (!inputFile.is_open()) {
        std::cerr << "Error: Cannot open file " << fileName << std::endl;
        return GraphBasedRepresentation();
    }
    
    int numberOfNodes;
    inputFile >> numberOfNodes;
    
    GraphBasedRepresentation graphTree(numberOfNodes);
    
    // Find root node (node that is not a child of any other node)
    std::vector<bool> isChild(numberOfNodes, false);
    
    for (int nodeIndex = 0; nodeIndex < numberOfNodes; ++nodeIndex) {
        int degreeCount;
        inputFile >> degreeCount;
        graphTree.adjacencyData[nodeIndex].resize(degreeCount);
        
        for (int childIndex = 0; childIndex < degreeCount; ++childIndex) {
            inputFile >> graphTree.adjacencyData[nodeIndex][childIndex];
            isChild[graphTree.adjacencyData[nodeIndex][childIndex]] = true;
        }
    }
    
    // Find root node
    for (int nodeIndex = 0; nodeIndex < numberOfNodes; ++nodeIndex) {
        if (!isChild[nodeIndex]) {
            graphTree.rootNode = nodeIndex;
            break;
        }
    }
    
    inputFile.close();
    return graphTree;
}

/**
 * @brief Display graph-based representation to console
 * @param graphTree The graph-based representation to display
 */
void displayGraphBasedRepresentation(const GraphBasedRepresentation& graphTree) {
    std::cout << "=== Graph-Based Representation (Adjacency List) ===" << std::endl;
    std::cout << "Number of nodes: " << graphTree.numberOfNodes << std::endl;
    std::cout << "Root node: " << graphTree.rootNode << std::endl;
    
    for (int nodeIndex = 0; nodeIndex < graphTree.numberOfNodes; ++nodeIndex) {
        std::cout << nodeIndex << ": ";
        if (graphTree.adjacencyData[nodeIndex].empty()) {
            std::cout << "(no children)";
        } else {
            for (int childIndex = 0; childIndex < static_cast<int>(graphTree.adjacencyData[nodeIndex].size()); ++childIndex) {
                std::cout << graphTree.adjacencyData[nodeIndex][childIndex];
                if (childIndex < static_cast<int>(graphTree.adjacencyData[nodeIndex].size()) - 1) {
                    std::cout << " ";
                }
            }
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

/**
 * @brief Write graph-based representation to output file
 * @param graphTree The graph-based representation to write
 * @param fileName Name of output file
 */
void writeGraphBasedRepresentationToFile(const GraphBasedRepresentation& graphTree, const std::string& fileName) {
    std::ofstream outputFile(fileName);
    if (!outputFile.is_open()) {
        std::cerr << "Error: Cannot create file " << fileName << std::endl;
        return;
    }
    
    outputFile << "=== Graph-Based Representation (Adjacency List) ===" << std::endl;
    outputFile << "Number of nodes: " << graphTree.numberOfNodes << std::endl;
    outputFile << "Root node: " << graphTree.rootNode << std::endl;
    
    for (int nodeIndex = 0; nodeIndex < graphTree.numberOfNodes; ++nodeIndex) {
        outputFile << nodeIndex << ": ";
        if (graphTree.adjacencyData[nodeIndex].empty()) {
            outputFile << "(no children)";
        } else {
            for (int childIndex = 0; childIndex < static_cast<int>(graphTree.adjacencyData[nodeIndex].size()); ++childIndex) {
                outputFile << graphTree.adjacencyData[nodeIndex][childIndex];
                if (childIndex < static_cast<int>(graphTree.adjacencyData[nodeIndex].size()) - 1) {
                    outputFile << " ";
                }
            }
        }
        outputFile << std::endl;
    }
    
    outputFile.flush();
    outputFile.close();
}

/**
 * @brief Convert array of parents to graph-based representation
 * @param parentArray Array of parent pointers
 * @param numberOfNodes Number of nodes in the tree
 * @return GraphBasedRepresentation structure containing the converted data
 */
GraphBasedRepresentation convertArrayParentsToGraphBased(const std::vector<int>& parentArray, 
                                                        int numberOfNodes) {
    GraphBasedRepresentation graphTree(numberOfNodes);
    
    for (int nodeIndex = 0; nodeIndex < numberOfNodes; ++nodeIndex) {
        if (parentArray[nodeIndex] != NIL_VALUE) {
            graphTree.adjacencyData[parentArray[nodeIndex]].push_back(nodeIndex);
        } else {
            graphTree.rootNode = nodeIndex;
        }
    }
    
    return graphTree;
}

/**
 * @brief Convert first-child next-sibling to graph-based representation
 * @param firstChildArray Array of first child pointers
 * @param nextSiblingArray Array of next sibling pointers
 * @param numberOfNodes Number of nodes in the tree
 * @return GraphBasedRepresentation structure containing the converted data
 */
GraphBasedRepresentation convertFirstChildNextSiblingToGraphBased(const std::vector<int>& firstChildArray, 
                                                                 const std::vector<int>& nextSiblingArray, 
                                                                 int numberOfNodes) {
    GraphBasedRepresentation graphTree(numberOfNodes);
    
    // Find root node (node that is not a child of any other node)
    std::vector<bool> isChild(numberOfNodes, false);
    for (int nodeIndex = 0; nodeIndex < numberOfNodes; ++nodeIndex) {
        if (firstChildArray[nodeIndex] != NIL_VALUE) {
            // Mark first child as a child
            isChild[firstChildArray[nodeIndex]] = true;
            
            // Mark all siblings in the chain as children
            int currentSibling = firstChildArray[nodeIndex];
            while (currentSibling != NIL_VALUE) {
                isChild[currentSibling] = true;
                currentSibling = nextSiblingArray[currentSibling];
            }
        }
    }
    
    // Find root node
    for (int nodeIndex = 0; nodeIndex < numberOfNodes; ++nodeIndex) {
        if (!isChild[nodeIndex]) {
            graphTree.rootNode = nodeIndex;
            break;
        }
    }
    
    // Convert first-child next-sibling to adjacency list
    for (int nodeIndex = 0; nodeIndex < numberOfNodes; ++nodeIndex) {
        int currentChild = firstChildArray[nodeIndex];
        while (currentChild != NIL_VALUE) {
            graphTree.adjacencyData[nodeIndex].push_back(currentChild);
            currentChild = nextSiblingArray[currentChild];
        }
    }
    
    return graphTree;
} 

File: Project2/Baitoan4/tree/py/graph_based_representation.py
"""
Graph-Based Representation implementation for Tree representations
"""

from typing import List
from dataclasses import dataclass

NIL_VALUE = -1


@dataclass
class GraphBasedRepresentation:
    """Structure to represent graph-based representation for tree"""
    adjacency_data: List[List[int]]
    number_of_nodes: int
    root_node: int
    
    def __init__(self, number_of_nodes: int = 0):
        self.number_of_nodes = number_of_nodes
        self.root_node = NIL_VALUE
        self.adjacency_data = [[] for _ in range(number_of_nodes)]


def read_graph_based_representation_from_file(file_name: str) -> GraphBasedRepresentation:
    """
    Read graph-based representation from input file
    
    Args:
        file_name: Name of input file
        
    Returns:
        GraphBasedRepresentation structure containing the tree data
    """
    try:
        with open(file_name, 'r') as input_file:
            number_of_nodes = int(input_file.readline().strip())
            
            graph_tree = GraphBasedRepresentation(number_of_nodes)
            
            for node_index in range(number_of_nodes):
                line = input_file.readline().strip().split()
                if line and line[0] != '0':
                    num_children = int(line[0])
                    for i in range(1, num_children + 1):
                        if i < len(line):
                            child = int(line[i])
                            graph_tree.adjacency_data[node_index].append(child)
            
            # Find root node (node with no incoming edges)
            has_incoming = [False] * number_of_nodes
            for node_index in range(number_of_nodes):
                for child in graph_tree.adjacency_data[node_index]:
                    has_incoming[child] = True
            
            for node_index in range(number_of_nodes):
                if not has_incoming[node_index]:
                    graph_tree.root_node = node_index
                    break
            
            return graph_tree
    except FileNotFoundError:
        print(f"Error: Cannot open file {file_name}")
        return GraphBasedRepresentation()
    except Exception as e:
        print(f"Error reading file: {e}")
        return GraphBasedRepresentation()


def display_graph_based_representation(graph_tree: GraphBasedRepresentation) -> None:
    """
    Display graph-based representation to console
    
    Args:
        graph_tree: The graph-based representation tree to display
    """
    print("=== Graph-Based Representation ===")
    print(f"Number of nodes: {graph_tree.number_of_nodes}")
    print(f"Root node: {graph_tree.root_node}")
    
    print("Adjacency lists:")
    for node_index in range(graph_tree.number_of_nodes):
        if graph_tree.adjacency_data[node_index]:
            children = " ".join(map(str, graph_tree.adjacency_data[node_index]))
            print(f"Node {node_index}: children = [{children}]")
        else:
            print(f"Node {node_index}: children = []")
    print()


def write_graph_based_representation_to_file(graph_tree: GraphBasedRepresentation, file_name: str) -> None:
    """
    Write graph-based representation to output file
    
    Args:
        graph_tree: The graph-based representation tree to write
        file_name: Name of output file
    """
    try:
        with open(file_name, 'w') as output_file:
            output_file.write("=== Graph-Based Representation ===\n")
            output_file.write(f"Number of nodes: {graph_tree.number_of_nodes}\n")
            output_file.write(f"Root node: {graph_tree.root_node}\n")
            
            output_file.write("Adjacency lists:\n")
            for node_index in range(graph_tree.number_of_nodes):
                if graph_tree.adjacency_data[node_index]:
                    children = " ".join(map(str, graph_tree.adjacency_data[node_index]))
                    output_file.write(f"Node {node_index}: children = [{children}]\n")
                else:
                    output_file.write(f"Node {node_index}: children = []\n")
    except Exception as e:
        print(f"Error: Cannot create file {file_name}: {e}")


def convert_graph_based_to_array_parents(adjacency_data: List[List[int]], 
                                        number_of_nodes: int) -> 'ArrayOfParents':
    """
    Convert graph-based representation to array of parents
    
    Args:
        adjacency_data: Adjacency list data (parent -> children)
        number_of_nodes: Number of nodes in the tree
        
    Returns:
        ArrayOfParents structure containing the converted data
    """
    from array_of_parents import ArrayOfParents
    
    array_parents = ArrayOfParents(number_of_nodes)
    
    # Find root and build parent relationships
    for parent_index in range(number_of_nodes):
        for child in adjacency_data[parent_index]:
            if 0 <= child < number_of_nodes:
                array_parents.parent_array[child] = parent_index
    
    # Find root node (node with no parent)
    for node_index in range(number_of_nodes):
        if array_parents.parent_array[node_index] == NIL_VALUE:
            array_parents.root_node = node_index
            break
    
    return array_parents


def convert_graph_based_to_first_child_next_sibling(adjacency_data: List[List[int]], 
                                                   number_of_nodes: int) -> 'FirstChildNextSibling':
    """
    Convert graph-based representation to first-child next-sibling
    
    Args:
        adjacency_data: Adjacency list data (parent -> children)
        number_of_nodes: Number of nodes in the tree
        
    Returns:
        FirstChildNextSibling structure containing the converted data
    """
    from first_child_next_sibling import FirstChildNextSibling
    
    fcns_tree = FirstChildNextSibling(number_of_nodes)
    
    # Find root node (node with no incoming edges)
    has_incoming = [False] * number_of_nodes
    for node_index in range(number_of_nodes):
        for child in adjacency_data[node_index]:
            has_incoming[child] = True
    
    for node_index in range(number_of_nodes):
        if not has_incoming[node_index]:
            fcns_tree.root_node = node_index
            break
    
    # Convert adjacency lists to first-child next-sibling
    for parent_index in range(number_of_nodes):
        children = adjacency_data[parent_index]
        if children:
            # Set first child
            fcns_tree.first_child_array[parent_index] = children[0]
            
            # Set next sibling relationships
            for i in range(len(children) - 1):
                fcns_tree.next_sibling_array[children[i]] = children[i + 1]
    
    return fcns_tree 